compiler/rustc_mir_transform/src/deref_separator.rs

```diff
@@ -8,11 +8,15 @@ pub fn deref_finder<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
     let mut patch = MirPatch::new(body);
     let (basic_blocks, local_decl) = body.basic_blocks_and_local_decls_mut();
     for (block, data) in basic_blocks.iter_enumerated_mut() {
+        let statement_len = data.statements.len();
         for (i, stmt) in data.statements.iter_mut().enumerate() {
             match stmt.kind {
                 StatementKind::Assign(box (og_place, Rvalue::Ref(region, borrow_knd, place))) => {
                     for (idx, (p_ref, p_elem)) in place.iter_projections().enumerate() {
-                        if p_elem == ProjectionElem::Deref && !p_ref.projection.is_empty() {
+                        if p_elem == ProjectionElem::Deref
+                            && !p_ref.projection.is_empty()
+                            && region.is_erased()
```

Comment:
still unsure about the reason for ICE having a tighter check avoids it.

Reply:
One problem may be that this opt is not sound yet when there is more than one deref in the middle of the projection. All projections are always off the base projection. This is fine for a first round, but you need to break from the projection iteration loop when you successfully apply your opt for now.

Reply:
I think that was it, you are right.